### PR TITLE
[c10d] added docstrings and tests for src / dst

### PR DESCRIPTION
**Pull request by weifengpy**
Follow up https://github.com/pytorch/pytorch/pull/118359: whether``src`` and ``dst`` are base on global pg or sub pg
* update c10d docstring: ``src`` / ``dst`` are base on global pg regardless of ``group`` arguments
* communication ops with ``dst`` argument: ``reduce``, ``gather_object``, ``gather``, ``send``, ``isend``
* communication ops with ``src`` argument: ``irecv``, ``recv``, ``broadcast``, ``broadcast_object_list``, ``scatter``, ``scatter_object_list``
* ``pytest test/distributed/test_c10d_nccl.py -k subgroup``

3 collectives are for pickable objects (``gather_object``, ``broadcast_object_list``, ``scatter_object_list``). There are 2 ways to set device
* use device argument: it's implemented in ``broadcast_object_list``. maybe worth implementing in the other 2
* ``torch.cuda.set_device(global_rank)``

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: test/distributed/test_c10d_nccl.py [hunk 1]
```diff
@@ -3532,6 +3537,15 @@ def test_new_group_local_sync_sanity_check(self):
     def test_new_group_local_sync_duplicated_pg(self):
         self._test_new_group_local_sync_duplicate_pg(backend="nccl")
 
+    def _init_pg_subgroups(self, world_size: int):
```
Comment (wconstab) [inline test/distributed/test_c10d_nccl.py:3540]: nit: its a little bad to expose an 'int' arg for world_size, but then also assume it's gonna be 4 (hardcoding 0,1 and 2,3).

suggest either naming the function about '2x2 subgroups' and then hardcode/assert 4, or, put the logic in here to take world_size and make 2 groups of [:size/2] and [size/2:] ranks.

(even if you make the helper parametrized on world_size, you can still leave the tests themselves hardcoded to 4 for simplicity)

File: test/distributed/test_c10d_nccl.py [hunk 2]
```diff
@@ -3589,6 +3592,138 @@ def test_gather_object_subgroup(self):
         else:
             torch.distributed.gather_object(input, object_gather_list=None, dst=self.rank - 1, group=my_group)
 
+    @requires_nccl()
+    @skip_if_lt_x_gpu(4)
+    def test_reduce_subgroup(self):
+        world_size = 4
+        if self.rank >= world_size:
+            return
+        group = self._init_pg_subgroups(world_size)
```
Comment (wconstab) [inline test/distributed/test_c10d_nccl.py:3601]: nit: name this subgroup?

**Review by wconstab — APPROVED** [review]
nice! thanks for following up on this.

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -1688,7 +1688,7 @@ def isend(tensor: torch.Tensor, dst: int, group: Optional[ProcessGroup] = None,
 
     Args:
         tensor (Tensor): Tensor to send.
-        dst (int): Destination rank.
+        dst (int): Destination rank on global process group (regardless of 'group' argument)
```
Comment (awgu) [inline torch/distributed/distributed_c10d.py:1691]: Screen Shot 2024-01-29 at 8 46 36 PM
nit: We can use \`\`group\`\` if we want `group` in code type.

File: test/distributed/test_c10d_nccl.py (hunk 1 above)
Comment (weifengpy) [reply test/distributed/test_c10d_nccl.py:3540]: good point. changed to ``_init_two_pg2_subgroups`` and assert on ``world_size==4``

File: torch/distributed/distributed_c10d.py (hunk 3 above)
Comment (weifengpy) [reply torch/distributed/distributed_c10d.py:1691]: will use \`\`group\`\` for better looking

**Pull request closed without merging**